src/mds/MDCache.cc

```diff
@@ -5524,8 +5524,15 @@ void MDCache::export_remaining_imported_caps()
     }
   }
 
+  for (map<inodeno_t, list<MDSInternalContextBase*> >::iterator p = cap_reconnect_waiters.begin();
+       p != cap_reconnect_waiters.end();
+       ++p)
+    mds->queue_waiters(p->second);
```

Comment:
In that case can we put it somewhere less buried?

And perhaps we should print out central log warnings for any unhandled cap flushes rather than just blindly applying them? Not sure, but I generally don't like quiet cleanups of bugs like this.